src/halmos/sevm.py

```diff
@@ -861,19 +862,15 @@ def returndatasize(self) -> int:
         return len(returndata) if returndata is not None else 0
 
     def is_jumpdest(self, x: Word) -> bool:
-        if not is_concrete(x):
-            return False
-
-        pc: int = int_of(x)
-        if pc < 0:
-            raise ValueError(pc)
-
+        pc = unbox_int(x)
         return pc in self.pgm.valid_jump_destinations()
 
     def jumpi_id(self) -> str:
-        return f"{self.pc}:" + ",".join(
-            map(lambda x: str(x) if self.is_jumpdest(x) else "", self.st.stack)
-        )
+        # no need to scan the entire stack, limit ourselves to the top N elements
+        # limiting ourselves to the top 16 leads to false positives
+        top_stack = self.st.stack[:32]
+        jumpdests_str = map(lambda x: str(x) if self.is_jumpdest(x) else "", top_stack)
+        return f"{self.pc}:{','.join(jumpdests_str)}"
```

Comment:
> could you elaborate why we don't need to scan the entire stack?

this is a costly operation, and since the loop detection is a heuristic I am looking for a good trade-off. I can try to think of a more reliable alternative

Reply:
yes i agree it is costly. and yes it is heuristic anyway, but new heuristic requires another validation process / period, which deserves its own pr and separate discussion.

a possible optimization without changing heuristic is to maintain the jumpi id inside the stack, but this could be nontrivial to implement.

i'd suggest not spending too much time on this loop detection heuristic, as we might want to develop a different loop detection, especially for the new branching condition solving.

Reply:
good point, I'll revert this particular change

Reply:
back to whole stack scanning in eac7662